Make Task#source default to the first prerequisite in non-rule tasks.

avdi opened the pull request:
In GNU Make, The `$<` macro always expands to the _first_ prerequisite of the current rule. This supports a common idiom wherein the primary "source file" for the target is listed first in the dependency list.

The Rake analog of this macro seems to be `Task#source`, but unlike in Make it is only available in tasks declared with `rule`.

Making `Task#source` return the first prerequisite in non-rule tasks aids code malleability. Let's say I start with a file task:

``` ruby
file "ch1.html" => ["ch1.md"] do |t|
  sh "pandoc -o #{t.name} #{t.source}"
end
```

I can then generalize it into a rule without changing the body at all:

``` ruby
rule ".html" => [".md"] do |t|
  sh "pandoc -o #{t.name} #{t.source}"
end
```

What do you think?

Commit:
Make Task#source default to the first prerequisite in non-rule tasks.

jimweirich commented:
Given that this makes "source" and alias for the first prerequisite, should we also make "sources" an alias for "prerequisites"?

avdi commented:
That seems reasonable at first glance. Would you like an amended PR?

jimweirich commented:
yes please.

Commit:
Make #sources a synonym for #prerequisites if @sources isn't set.

I've opted for a pretty conservative definition of this - `#sources`
will prefer to return `@sources` even if `@sources` is set to `nil`

avdi commented:
OK, done. I've opted for a pretty conservative definition of this - `#sources` will prefer to return `@sources` even if `@sources` is set to `nil`

avdi commented:
It might be preferable to use `@sources || prerequisites` though... thoughts?

avdi commented:
Now to to submit an equivalent of Make's "all sources newer than target" special variable...

dsisnero commented:
I don't think this is a good idea.  Say I want to make a task that copies files (stealing ideas from gradle)

``` ruby
copy do
  source = FileList('lib/'**./*)
  target = "install_dir"
  expand( copyright: 'MIT', version: 1)
end
```

But I guess I could just redefine it in the copy task.

jimweirich commented:
dsisnero: Can you clarify your message.  What is a bad idea? Making Task#source default to preferences, or adding an "all sources newer than target" special variable (from the comment immediately before yours?  

And I'm not quite sure what you believe to be a problem (the "source" in your code snippet is a local variable, I'm not sure the code is doing what you think).
